### PR TITLE
Entropy calculation of sections and overlays (PE, ELF, COFF, MACHO)

### PR DESCRIPTION
**Pull request by JakubPruzinec**
https://github.com/avast-tl/retdec/issues/502
I've implemented extraction of section and overlay entropies. Algorithm for entropy calculation uses Shannons formula normalized to ~<0,1>~ <0,8> interval.

### CODE REVIEW

File: src/fileformat/types/sec_seg/sec_seg.cpp [hunk 1]
```diff
@@ -521,6 +536,27 @@ void SecSeg::setMemory(bool sMemory)
...
+ * Compute entropy of section data in <0,1>
+ */
+void SecSeg::computeEntropy()
+{
+	if (!loaded)
+	{
+		return;
+	}
+	
+	auto data = reinterpret_cast<const uint8_t *>(bytes.data());
```
Comment (metthal) [inline src/fileformat/types/sec_seg/sec_seg.cpp]: Is `reinterpret_cast` really necessary here? Wouldn't `static_cast` be enough or is the cast necessary at all?

File: src/fileinfo/file_wrapper/pe/pe_wrapper.cpp [hunk 2]
```diff
@@ -176,9 +176,15 @@ bool PeWrapper::getFileSection(unsigned long long secIndex, FileSection &section
...
 		{
+			double entropy;
+			if(auxSec->getEntropy(entropy))
+			{
+				section.setEntropy(entropy);
+			}
 			section.setCrc32(auxSec->getCrc32());
 			section.setMd5(auxSec->getMd5());
 			section.setSha256(auxSec->getSha256());
+
```
Comment (metthal) [inline src/fileinfo/file_wrapper/pe/pe_wrapper.cpp]: Remove this unnecessary whitespace please.

File: src/fileformat/types/sec_seg/sec_seg.cpp (hunk 1 above)
Comment (JakubPruzinec) [reply src/fileformat/types/sec_seg/sec_seg.cpp]: Cast is necessary, since `bytes.data()` returns `const char *` and `computeDataEntropy` takes `const unsigned char*` as parameter. As stated earlier, `static_cast` doesn't suffice. `invalid static_cast from type ‘const char*’ to type ‘const uint8_t*’ {aka ‘const unsigned char*’}`, please see https://stackoverflow.com/questions/658913/c-style-cast-from-unsigned-char-to-const-char

Comment (JakubPruzinec) [reply src/fileformat/types/sec_seg/sec_seg.cpp]: Also, I wouln't be so concerned about wrong interpretation of signedness as `computeDataEntropy` doesn't take it into consideration.

**Comment by metthal** [comment]
Could you please also modify it so that results are in range from 0 to 8? We should stick with what [VirusTotal](https://www.virustotal.com/#/file/1054ab5902dd121ab49ae855415926bbc5b5ed974444cc8543a7b9388a82a098/details) and [YARA](https://github.com/VirusTotal/yara/blob/master/libyara/modules/math.c#L68) are using.

File: src/fileformat/utils/other.cpp [hunk 3]
```diff
@@ -99,5 +101,38 @@ std::vector<std::string> getSupportedArchitectures()
...
+	{
+		histogram[data[i]]++;
+	}
+
+	for (auto frequency : histogram)
+	{
+		if (frequency)
+		{
+			double probability = static_cast<double>(frequency) / dataLen;
+			entropy -= probability * (log(probability) / log(2));
```
Comment (metthal) [inline src/fileformat/utils/other.cpp]: This can be calculated using only single logarithm because `log_a(x) / log_a(b) = log_b(x)` so you can actually use logarithm of base-2 here. Also please use `std::` namespace instead of the functions in the global namespace.

File: src/fileformat/utils/other.cpp [hunk 4]
```diff
@@ -4,8 +4,10 @@
  * @copyright (c) 2017 Avast Software, licensed under the MIT license
  */
 
+#include <array>
 #include <map>
 #include <unordered_map>
+#include <cmath>
```
Comment (metthal) [inline src/fileformat/utils/other.cpp]: Please keep headers in alphabetical order.

File: include/retdec/fileformat/types/sec_seg/sec_seg.h [hunk 5]
```diff
@@ -47,10 +47,12 @@ class SecSeg
 		unsigned long long address;       ///< start address in memory
 		unsigned long long memorySize;    ///< size in memory
 		unsigned long long entrySize;     ///< size of one entry in file
+		double entropy;                   ///< entropy in <0,1>
```
Comment (metthal) [inline include/retdec/fileformat/types/sec_seg/sec_seg.h]: Isn't comment here wrong? Shouldn't it be `<0,8>`?
